### PR TITLE
Store Event in Datastore

### PR DESCRIPTION
**Pull request by mbouadeus**
Create activity servlet creates activity event and stores it in datastore.

### CODE REVIEW

File: src/main/java/com/google/sps/data/Activity.java [hunk 1]
```diff
@@ -1,20 +1,37 @@
...
 
+import com.google.appengine.api.datastore.Key;
+
 public class Activity {
+  private Key id;
   private String name;
   private Category category;
   private String url;
 
+
```
Comment (bhargaviks) [inline src/main/java/com/google/sps/data/Activity.java:11]: remove extra lines

File: src/main/java/com/google/sps/data/Activity.java [hunk 2]
```diff
@@ -1,20 +1,37 @@
...
+  }
+  
+  public String getName() {
+    return name;
+  }
+
+  public String getUrl() {
+    return url;
+  }
+
```
Comment (bhargaviks) [inline src/main/java/com/google/sps/data/Activity.java:39]: add new line at EOF

File: src/main/java/com/google/sps/data/ActivityEvent.java [hunk 3]
```diff
@@ -2,16 +2,38 @@
 
 import java.util.List;
 
-public class ActivityEvent {
+public final class ActivityEvent {
```
Comment (bhargaviks) [inline src/main/java/com/google/sps/data/ActivityEvent.java:5]: why do we need final here? Is it necessary?

File: src/main/java/com/google/sps/data/ActivityEvent.java [hunk 4]
```diff
@@ -2,16 +2,38 @@
...
+    return endTimestamp;
+  }
+
+  public Activity getActivity() {
+    return activity;
+  }
+
+  public List<String> getGuests() {
+    return guests;
+  }
```
Comment (bhargaviks) [inline src/main/java/com/google/sps/data/ActivityEvent.java:45]: Add new line at EOF

File: src/main/java/com/google/sps/data/EventUtility.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    activityMap.put("id", activityEntity.getKey());
+
+    return activityMap;
+  }
+
+  /*
+  * Store activity event into datastore.
+  */
+  public static void storeActivityEvent(ActivityEvent event) {
+    // Prepare database
```
Comment (bhargaviks) [inline src/main/java/com/google/sps/data/EventUtility.java]: add punctuation

File: src/main/java/com/google/sps/data/EventUtility.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Entity eventEntity = new Entity(eventCollectionID);
+
+    // Get all activity event properties
```
Comment (bhargaviks) [inline src/main/java/com/google/sps/data/EventUtility.java]: add punctuation

File: src/main/java/com/google/sps/data/EventUtility.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    String userId = event.getUserId();
+    String startTimestamp = new String(event.getStartTimestamp());
+    String endTimestamp = new String(event.getEndTimestamp());
+    String activityId = event.getActivity().getId();
+    String guests = String.join(",", event.getGuests());
+
+    // Set the properties into the event entity
```
Comment (bhargaviks) [inline src/main/java/com/google/sps/data/EventUtility.java]: add punctuation.

File: src/main/java/com/google/sps/data/ActivityEvent.java [hunk 8]
```diff
@@ -2,16 +2,38 @@
...
+
+  public long getStartTimestamp() {
+    return startTimestamp;
+  }
+
+  public long getEndTimestamp() {
+    return endTimestamp;
+  }
+
+  public Activity getActivity() {
```
Comment (bhargaviks) [inline src/main/java/com/google/sps/data/ActivityEvent.java:39]: All these functions that return can have some javadoc that tells us what is being returned and what that's used for. Similar to how TimeRange was made in week5-tdd.

File: src/main/java/com/google/sps/data/EventUtility.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
+package com.google.sps.data;
+
```
Comment (bhargaviks) [inline src/main/java/com/google/sps/data/EventUtility.java:2]: Always make sure to add a new line at all files at EOF

File: src/main/java/com/google/sps/data/GeneralUtility.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  * was not specified by the client
+  */
+  public static String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null) {
+      return defaultValue;
+    }
+    return value;
+  }
+}
```
Comment (bhargaviks) [inline src/main/java/com/google/sps/data/GeneralUtility.java:24]: add new line at EOF

File: src/main/java/com/google/sps/servlets/CreateEventServlet.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    String[] paramNames = {"userID", "accessToken", "startTimestamp", "endTimestamp", "activityKey", "guests"};
+
+    for (String paramName : paramNames) {
+      String paramValue = getParameter(paramName);
+      eventInfo.put(paramName, paramValue);
+    }
+
+    return eventInfo;
+  }
+}
```
Comment (bhargaviks) [inline src/main/java/com/google/sps/servlets/CreateEventServlet.java:51]: add new line at EOF

File: src/main/java/com/google/sps/servlets/RetrieveActivitiesServlet.java [hunk 12]
```diff
@@ -11,6 +11,8 @@
 import com.google.sps.data.Activity;
 import com.google.sps.data.Activity.Category;
 import static com.google.sps.data.ActivitiesUtility.getActivities;
+import static com.google.sps.data.GneralUtility.getParameter;
+
```
Comment (bhargaviks) [inline src/main/java/com/google/sps/servlets/RetrieveActivitiesServlet.java]: remove extra line.

Comment (bhargaviks) [reply src/main/java/com/google/sps/servlets/RetrieveActivitiesServlet.java]: In the same servlet file, there are extra lines at lines 32,33,34. Make sure to remove the two extra lines.

File: src/main/java/com/google/sps/servlets/RetrieveActivitiesServlet.java [hunk 13]
```diff
@@ -57,16 +59,4 @@ private static Category getActivityCategory(HttpServletRequest request) {
...
-  * was not specified by the client
-  */
-  private static String getParameter(HttpServletRequest request, String name, String defaultValue) {
-    String value = request.getParameter(name);
-    if (value == null) {
-      return defaultValue;
-    }
-    return value;
-  }
 }
```
Comment (bhargaviks) [inline src/main/java/com/google/sps/servlets/RetrieveActivitiesServlet.java]: add new line at EOF

**Review by bhargaviks — APPROVED** [review]

File: src/main/java/com/google/sps/data/ActivityEvent.java (hunk 3 above)
Comment (mbouadeus) [reply src/main/java/com/google/sps/data/ActivityEvent.java:5]: It's not necessary. It's just to say that the class can't be extended. It's good semantics though.

File: src/main/java/com/google/sps/data/Activity.java (hunk 2 above)
Comment (mbouadeus) [reply src/main/java/com/google/sps/data/Activity.java:39]: There is a new line.

Comment (mbouadeus) [reply src/main/java/com/google/sps/data/Activity.java:39]: You would see a red mark at the end of the file on github if there wasn't.

File: src/main/java/com/google/sps/data/ActivityEvent.java (hunk 4 above)
Comment (mbouadeus) [reply src/main/java/com/google/sps/data/ActivityEvent.java:45]: There's a new line already.

File: src/main/java/com/google/sps/data/GeneralUtility.java (hunk 10 above)
Comment (mbouadeus) [reply src/main/java/com/google/sps/data/GeneralUtility.java:24]: There's a new line already.

File: src/main/java/com/google/sps/servlets/CreateEventServlet.java (hunk 11 above)
Comment (mbouadeus) [reply src/main/java/com/google/sps/servlets/CreateEventServlet.java:51]: There's a new line already.

File: src/main/java/com/google/sps/servlets/RetrieveActivitiesServlet.java (hunk 12 above)
Comment (mbouadeus) [reply src/main/java/com/google/sps/servlets/RetrieveActivitiesServlet.java]: This servlet is acutally no longer necessary. I'll go ahead and delete it.

File: src/main/java/com/google/sps/servlets/RetrieveActivitiesServlet.java (hunk 13 above)
Comment (mbouadeus) [reply src/main/java/com/google/sps/servlets/RetrieveActivitiesServlet.java]: Removing file.

File: src/main/java/com/google/sps/data/Activity.java (hunk 1 above)
Comment (mbouadeus) [reply src/main/java/com/google/sps/data/Activity.java:11]: Fixed it.

File: src/main/java/com/google/sps/data/EventUtility.java (hunk 5 above)
Comment (mbouadeus) [reply src/main/java/com/google/sps/data/EventUtility.java]: Fixed it.

File: src/main/java/com/google/sps/servlets/RetrieveActivitiesServlet.java (hunk 12 above)
Comment (mbouadeus) [reply src/main/java/com/google/sps/servlets/RetrieveActivitiesServlet.java]: Fixed it.

File: src/main/java/com/google/sps/servlets/RetrieveActivitiesServlet.java (hunk 13 above)
Comment (mbouadeus) [reply src/main/java/com/google/sps/servlets/RetrieveActivitiesServlet.java]: Fixed it.

File: src/main/java/com/google/sps/data/Activity.java (hunk 2 above)
Comment (bhargaviks) [reply src/main/java/com/google/sps/data/Activity.java:39]: That's what I always thought too. But when I view the file, I don't see the end of line. :( BUt good, if it's there then you're good.

File: src/main/java/com/google/sps/data/Activity.java [hunk 14]
```diff
@@ -9,12 +12,25 @@
...
 
+  public Key getId() {
+    return id;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public String getUrl() {
```
Comment (pengxianzhe) [inline src/main/java/com/google/sps/data/Activity.java:36]: You are missing the getter method for category, is this intended?

File: src/main/java/com/google/sps/data/EventUtility.java [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+  * Store activity event into datastore.
+  */
+  public static void storeActivityEvent(ActivityEvent event) {
+    // Prepare database.
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Entity eventEntity = new Entity(eventCollectionID);
+
+    // Get all activity event properties.
+    String userId = event.getUserId();
```
Comment (pengxianzhe) [inline src/main/java/com/google/sps/data/EventUtility.java:42]: nit: you don't need to declare these variables. Putting them together with setProperty is fine.

File: src/main/java/com/google/sps/servlets/CreateEventServlet.java [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+
+    // Get event information
+    Map<String, String> eventInfo = getParameters(request);
+
+    ActivityEvent event = getActivityEvent(eventInfo);
+
+    // Store activity event in datastore
+    storeActivityEvnet(event);
+
+    // Create calendar event and send invite to guests
```
Comment (trinitydonohugh) [inline src/main/java/com/google/sps/servlets/CreateEventServlet.java]: This should probably be a "TODO" comment?

**Review by trinitydonohugh — APPROVED** [review]

File: src/main/java/com/google/sps/data/EventUtility.java [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+
+    return event;
+  }
+
+  /*
+  * Creates activity object using activity information retreived using
+  * the datastore activity id.
+  */
+  private static Activity getActivity(String activityId) {
+    Map<String, String> activityInfo = getActivityInfo(activityId);
```
Comment (pengxianzhe) [inline src/main/java/com/google/sps/data/EventUtility.java]: I start to wonder if this Map is really needed. It looks like you can just return the Entity here and create Activity object using the entity.

File: src/main/java/com/google/sps/servlets/CreateEventServlet.java (hunk 16 above)
Comment (mbouadeus) [reply src/main/java/com/google/sps/servlets/CreateEventServlet.java]: yes, your right.

File: src/main/java/com/google/sps/data/ActivityEvent.java (hunk 3 above)
Comment (isaacwdavis) [reply src/main/java/com/google/sps/data/ActivityEvent.java:5]: +1 to this being a good practice.

**Review by isaacwdavis — APPROVED** [review]

**Pull request merged**